### PR TITLE
13+ Field office support and single toon combos

### PR DESCRIPTION
**Pull request by Stoxis**
I've added support for cogs level 13 and over and changed it so single toon combos are also displayed.

### CODE REVIEW

**Comment by zzzachzzz** [comment]
Thank you for the PR! I will test this out tomorrow

**Comment by Stoxis** [comment]
Okay, if you need any updates to the PR feel free to ask! 👍

File: index.js [hunk 1]
```diff
@@ -89,15 +98,15 @@ function CogLvlCell(cogLvl) {
 function CogLvlColumn() {
   let arr = [];
   // TODO For lvl 13+ cogs
-  for (let cogLvl = 12; cogLvl >= 1; cogLvl--) {
+  for (let cogLvl = 20; cogLvl >= 1; cogLvl--) {
```
Comment (zzzachzzz) [inline index.js:101]: This is where we'd conditionally specify iterating through levels 20 to 1 for TTR, and levels 12 to 1 for classic Toontown. For example:
```suggestion
  const maxCogLvl = state.game === 'ttr' ? 20 : 12;
  for (let cogLvl = maxCogLvl; cogLvl >= 1; cogLvl--) {
```

File: index.js [hunk 2]
```diff
@@ -89,15 +98,15 @@ function CogLvlCell(cogLvl) {
...
     arr.push(CogLvlCell(cogLvl));
   }
   return arr.join('');
 }
 
 function CombosGrid() {
   const arr = [];
-  for (let cogLvl = 12; cogLvl >= 1; cogLvl--) {
+  for (let cogLvl = 20; cogLvl >= 1; cogLvl--) {
```
Comment (zzzachzzz) [inline index.js:109]: Same thing here. Maybe since it's used in multiple places, we specify this as a function somewhere:
```js
const getMaxCogLvl = () => state.game === 'ttr' ? 20 : 12;
```

File: index.js [hunk 3]
```diff
@@ -71,11 +76,15 @@ function Combo({ cogLvl, gagTrack, numToons, stunTrack }) {
 
 function CogLvlCell(cogLvl) {
   const cogHp = (state.game === 'ttr' ? ttrCogHp : classicCogHp)[cogLvl];
-
+  if (cogLvl > 12) {
+  	var cogLvlImg = "13+"
+  } else {
+  	var cogLvlImg = cogLvl
+  }
```
Comment (zzzachzzz) [inline index.js:83]: This would be a little cleaner:
```js
const cogLvlImg = cogLvl > 12 ? '13+' : cogLvl.toString();
```

File: gags.js [hunk 4]
```diff
@@ -244,8 +251,10 @@ function findCombo({
   game = 'ttr',
   stunTrack = null
 }) {
-  if (numToons === 1 && gagTrack === 'drop')
-    throw new Error('Invalid arguments: a drop combo must have 2 or more toons (for a stun gag)');
+  if (numToons === 1 && gagTrack === 'drop') {
+    // For drop combos with one toon, add two skips
+    numToons += 2;
```
Comment (zzzachzzz) [inline gags.js:256]: I'm a little confused on what this does? Since you included this:
```js
if (!(gagTrack === 'drop' && numToons === 1)) {
  arr.push(Combo({ numToons, cogLvl, gagTrack, stunTrack }));
}
```
it looks like we're still not calling `findCombo` with `gagTrack='drop'` and `numToons=1` specified.

Alternatively, for single toon drop combos, we could forego the stun gag, but I'm fine with the current solution of not displaying single toon drop combos.

**Review by zzzachzzz — CHANGES_REQUESTED** [review]
Thank you again for the PR! Really cool to see support for level 13+ cogs. Let me know if you have any questions on the feedback.

File: index.js (hunk 1 above)
Comment (Stoxis) [reply index.js]: change implemented.

File: index.js (hunk 2 above)
Comment (Stoxis) [reply index.js]: good idea, implemented.

File: gags.js [hunk 5]
```diff
@@ -11,12 +11,19 @@ export const classicCogHp = {
...
+  17: 356,
+  18: 394,
+  19: 434,
+  20: 476,  
 };
 
 export const ttrCogHp = {
   ...classicCogHp,
   12: 196,
-  // TODO 13+
```
[the remark this answers is not in the document]
Comment (Stoxis) [reply gags.js:19]: TTO Dropdown option now hides level 13+ cogs and ttr health values are in the correct array now.

File: index.js (hunk 3 above)
Comment (Stoxis) [reply index.js]: implemented.

File: gags.js (hunk 4 above)
Comment (Stoxis) [reply gags.js]: True, that's actually just an artifact, it was the first thing I changed in the code when I was still figuring out how everything was connected, I've undone this modification as it is pointless now due to upstream changes.

**Comment by Stoxis** [comment]
All suggested changes implemented, feel free to check out the new changes and tell me if you find them adequate.

**Comment by Stoxis** [comment]
I also noticed something weird is happening with the drop gags when you change between TTO and TTR but I'm not sure what is causing that.

**Comment by zzzachzzz** [comment]
I'll get to the follow up review when I can. I have the flu and have been feeling pretty lousy :(

**Comment by Stoxis** [comment]
> I'll get to the follow up review when I can. I have the flu and have been feeling pretty lousy :(

Oh, that sucks. No rush! Drink plenty of fluids and get plenty of rest, I'll still be here 👍 

Get well soon.

**Comment by Stoxis** [comment]
Noticed on page reload the checkbox could become desynchronized from `state.toggle` value, that's fixed now.

File: index.js [hunk 6]
```diff
@@ -4,7 +4,7 @@ import './index.css';
 const savedState = JSON.parse(localStorage.getItem('savedState'));
 
 const _selectedOrgGags =
-  savedState?.selectedOrgGags || Array.from({ length: 4 }, () => null);
+  savedState?.selectedOrgGags || Array.from({ length: 5 }, () => null);
```
Comment (zzzachzzz) [inline index.js:7]: Why was this changed from 4 to 5?

File: index.js [hunk 7]
```diff
@@ -18,6 +18,7 @@ const state = {
   selectedOrgGags: _selectedOrgGags,
   selectedOrgGagCounts: calcSelectedGagTrackCounts(_selectedOrgGags),
   isLured: savedState?.isLured || false,
+  toggle: savedState?.toggle || false,
```
Comment (zzzachzzz) [inline index.js:21]: Could this toggle be named something more descriptive, such as `hideLvl13UpCogs`?

**Review by zzzachzzz — COMMENTED** [review]
Looks great! Just 2 minor comments

File: index.js (hunk 6 above)
Comment (Stoxis) [reply index.js:7]: Oh, I thought it was related to the localStorage savedState code, but now that I look at it, it's about organic gags, I'll reverse that change.

File: index.js (hunk 7 above)
Comment (Stoxis) [reply index.js:21]: Yes, I'll do that.

**Comment by Stoxis** [comment]
One minor bug I noticed is since `grid-template: repeat(20, 140px) / repeat(6, 1fr);` and `grid-template-rows: repeat(20, 140px);` aren't updated to 12 when the toggle box is used, you can scroll the page pretty far down in 12 cog mode. Overriding the css would fix this but I'm not sure if you think it's worth the effort.

**Comment by Stoxis** [comment]
Never-mind, I added the dynamic CSS code.

**Comment by zzzachzzz** [comment]
> image
> This is not relevant to this PR but I suggest implementing a feature that detects mirrored combos like this and removing one of them as it's unnecessary to display a combo twice.

Is this a common occurrence? I think it's good to be explicit about that combo being the optimal choice for both a 3 and 4 toon combo

**Comment by Stoxis** [comment]
> > image
> > This is not relevant to this PR but I suggest implementing a feature that detects mirrored combos like this and removing one of them as it's unnecessary to display a combo twice.
> 
> Is this a common occurrence? I think it's good to be explicit about that combo being the optimal choice for both a 3 and 4 toon combo

image
For low(1-4) and high level(17-20) cogs it makes up the majority of the gag choices generated, I think it would lead to a much easier to navigate graph to hide duplicates, you could just hide 2 toon if 3 toon matches it and so on, you could also add an option to override this and display duplicates.

**Comment by zzzachzzz** [comment]
@Stoxis Interesting, and I agree that's out of scope for this PR.

I'll approve the current changes and I'll try to get around to deploying to the site within the next couple days.

**Comment by Stoxis** [comment]
That appears to only be the first commit.
